cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py

```diff
@@ -17,18 +17,293 @@
+                )
+            if self.dynamic_nodes > 0:
+                gres_render += (
+                    f"NodeName={self.dynamic_node_name()} Name=gpu "
+                    f"Type={ self.gpu_type } File=/dev/nvidia[0-{ self.gpu_count - 1 }]\n"
+                )
+
+        return gres_render
+
+    def _definitions(self, dynamic=False):
+        definitions = f" CPUs={ self.vcpus() } RealMemory={ self.real_memory } State=CLOUD {self._features(dynamic)}"
+
+        if self.has_gpu and self.gpu_count > 0:
+            definitions += f" Gres=gpu:{ self.gpu_type }:{self.gpu_count}"
+
+        return definitions
+
+    def _features(self, dynamic=False):
+        resource_type = "static"
+        if dynamic:
+            resource_type = "dynamic"
+
+        instance_type = f",{self.instance_types[0]}"
+        if len(self.instance_types) > 1:
+            # When multiple instance types are defined we do not know in advance which one will be used
+            # to launch the node. So we do not list any of them as feature
+            instance_type = ""
+
+        features = f"Feature={resource_type}{instance_type},{self.name}"
+        if self.efa_enabled:
+            features += ",efa"
+
+        if self.gpu_count > 0:
+            features += ",gpu"
+
+        return features
+
+    def _node_name(self, type, size):
+        return f"{self.queue_name}-{type}-{self.name}-[1-{ size }]"
+
+
+class QueueRenderer:
+    """Renders a PCluster Queue Config as a Slurm partition config or as gres config."""
+
+    def __init__(self, queue_config, no_gpu, memory_ratio, instance_types_info, conf_type="partition", default=False):
+        self.name = queue_config["Name"]
+        self.is_default = default
+        self.conf_type = conf_type
+        self.capacity_type = queue_config["CapacityType"]
+        self.iam = queue_config["Iam"]
+        self.compute_renderers = [
+            ComputeResourceRenderer(self.name, compute_resource_config, no_gpu, memory_ratio, instance_types_info)
+            for compute_resource_config in queue_config["ComputeResources"]
+        ]
+
+    def render(self):
+        """Launch the rendering of the required configuration."""
+        rendered = f"{CONFIG_HEADER}\n"
+        if self.conf_type == "gres":
+            rendered += self._as_gres()
+        else:
+            rendered += self._as_partition()
+
+        return rendered
+
+    def _as_partition(self):
+        partition_render = "\n"
+        for renderer in self.compute_renderers:
+            partition_render += f"{renderer.render()}"
+
+        partition_render += f"\n{self._render_node_set()}\n{self._render_partition()}\n"
+        return partition_render
+
+    def _as_gres(self):
+        gres_render = ""
+        for renderer in self.compute_renderers:
+            gres_render += f"{renderer.as_gres()}"
+
+        return gres_render
+
+    def _render_node_set(self):
```

Comment:
Nitpick: I believe we should use `nodeset` instead of `node_set` just because this way it's easier to find all info about nodesets with a single case-insensitive search (otherwise we would need a regex that may include the underscore).

Reply:
Good catch.
I agree and I'll fix it